### PR TITLE
Extend "full" config to non-boolean options and pass Clang+Asan

### LINKED ISSUE #2598: Havege does arithmetic on signed int
### Description
- Type: Bug
- Priority: Major

### Runtime failure with Asan

1. Check out https://github.com/ARMmbed/mbedtls at f790a6cbeea43ae297ef75dfd697e5cf30b94913 or https://github.com/ARMmbed/mbedtls/tree/mbedtls-2.16 at 20d707dd3eb9deadb5efe94c09fa6f105c56c5f1 or https://github.com/ARMmbed/mbed-crypto at 521dbc67dabf5f5dc2ad8148b6f01662805caac7.
2. Set the following configuration:  
    `scripts/config.pl full; scripts/config.pl unset MBEDTLS_MEMORY_BACKTRACE`  
    or
    `scripts/config.pl full; scripts/config.pl unset MBEDTLS_MEMORY_BACKTRACE; scripts/config.pl unset MBEDTLS_MEMORY_BUFFER_ALLOC_C; scripts/config.pl unset MBEDTLS_MEMORY_DEBUG`  
3. Build with **clang** and ASan (I used clang 3.8.0 Ubuntu 16.04). I've reproduced it with both in-tree and out-of-tree builds, both with `Asan` and `AsanDbg`.
    ```
    mkdir build-asan
    cd build-asan
    CC=clang cmake -D CMAKE_BUILD_TYPE:String=AsanDbg .
    make -j3
    cd tests
    ./test_suite_entropy
    ```
4. Expected: the tests pass. Actual: ASan runtime errors like the following (the exact values change at every run):
    ```
    /tmp/mbedtls$ ./test_suite_entropy
    Create NV seed_file ............................................... PASS
    Entropy write/update seed file .................................... /tmp/mbedtls/library/havege.c:179:9: runtime error: left shift of 1813511924 by 23 places cannot be represented in type 'int'
    SUMMARY: AddressSanitizer: undefined-behavior /tmp/mbedtls/library/havege.c:179:9 in 
    /tmp/mbedtls$ ./test_suite_entropy
    Create NV seed_file ............................................... PASS
    Entropy write/update seed file .................................... /tmp/mbedtls/library/havege.c:179:9: runtime error: left shift of negative value -1080004887
    SUMMARY: AddressSanitizer: undefined-behavior /tmp/mbedtls/library/havege.c:179:9 in 
    ```

Note that I'm getting Asan complaints only with Clang, not with GCC. We run this test (in the configuration `full` minus `MBEDTLS_MEMORY_BACKTRACE`) on CI, but only with GCC, not with Clang, which is why we didn't notice sooner. I'm still surprised we never noticed in developer builds, though. I often test with clang+asan! So there may still be some extra condition needed to reproduce the complaints.

### Analysis of `havege.c`

`havege.c` does indeed perform bit shifts on `int` that may cause an overflow. A shift that overflows a signed integer has undefined behavior in C. Asan's complaints are legitimate.

### LINKED ISSUE #2671: `config.pl full` or just turning on MBEDTLS_CHECK_PARAMS breaks the build
### Description
- Type: Bug
- Priority: Major

The `full` configuration breaks normal programs. We don't explicitly document this configuration, and it's intended for testing mbedtls rather than for production. I certainly wouldn't want to assert that this configuration is good for security, for example. Nonetheless, I would not expect this configuration to break perfectly valid programs so much that they don't compile.

Specifically, the problem is that `config.pl full` enables `MBEDTLS_CHECK_PARAMS`, but with the default setting of `MBEDTLS_PARAM_FAILED`, which calls a function (`mbedtls_param_failed`) that the user must supply. This is not a sensible default.

Justification for severity: breaks perfectly sensible code such as https://github.com/ARMmbed/mbedtls/pull/1622/commits/3cd7a32b373759dac3f47f0c2c434205be386810 (CI run [internal link]: https://jenkins-internal.mbed.com/job/mbedtls-pr-multibranch/job/PR-1622-head/4/execution/node/819/log).

Steps to reproduce (from an mbedtls checkout, e.g. at 7be9b4e9320597a8d2cf8162aa7ea4ee35f3c976 which is the head of `development` right now):
```
git submodule update
make clean
scripts/config.pl full
make
cat <<'EOF' >|a.c
#include <mbedtls/bignum.h>
int main(void)
{
    mbedtls_mpi x;
    mbedtls_mpi_init(&x);
    if (mbedtls_mpi_lset(&x, 42)) return 1;
    return 0;
}
EOF
gcc -O -Wall -I include -L library -L crypto/library a.c -lmbedcrypto
```

Expected behavior: successful build.

Actual behavior:
```
crypto/library/libmbedcrypto.a(bignum.o): In function `mbedtls_mpi_grow':
bignum.c:(.text+0x597): undefined reference to `mbedtls_param_failed'
crypto/library/libmbedcrypto.a(bignum.o): In function `mbedtls_mpi_shrink':
bignum.c:(.text+0x698): undefined reference to `mbedtls_param_failed'
crypto/library/libmbedcrypto.a(bignum.o): In function `mbedtls_mpi_copy':
bignum.c:(.text+0x7c6): undefined reference to `mbedtls_param_failed'
bignum.c:(.text+0x7e1): undefined reference to `mbedtls_param_failed'
crypto/library/libmbedcrypto.a(bignum.o): In function `mbedtls_mpi_safe_cond_assign':
bignum.c:(.text+0x991): undefined reference to `mbedtls_param_failed'
crypto/library/libmbedcrypto.a(bignum.o):bignum.c:(.text+0x9ac): more undefined references to `mbedtls_param_failed' follow
collect2: error: ld returned 1 exit status
```

I don't know what the best remedy is. We could simply keep `MBEDTLS_CHECK_PARAMS` off in the `full` config. I think `MBEDTLS_CHECK_PARAMS` is not useful, so this solution is attractive to me, but doing this breaks the expectation that `config.pl full` turns out everything that the library can do and that isn't contradictory with something else. Furthermore, beyond `config.pl full`, there is the problem that the default behavior of just turning on the option isn't sensible: I think the default implementation of `MBEDTLS_PARAM_FAILED`, if you just turn on `MBEDTLS_CHECK_PARAMS` without any other related configuration changes, should be `assert`.

### LINKED ISSUE #2682: make rebuilds programs
### Description
- Type: Bug
- Priority: Minor

---------------------------------------------------------------
## Bug

Since some submodule-related change, `make` rebuilds programs even when they don't need to be rebuilt.

**Expected behavior**   

After a successful build, running `make` again does nothing.

**Actual behavior**  

After a successful build, running `make` again rebuilds everything under `programs`.

**Steps to reproduce**  

```
make
make
```

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
* Tweak `config.h` so that `config.pl full` also enables the definition of `MBEDTLS_PARAM_FAILED` as `assert` (without this, all executables must have an `mbedtls_param_failed` function).
* Support `MBEDTLS_PARAM_FAILED` defined as `assert`. Together with the previous point, this fixes #2671.
* Define a `config.pl veryfull` that also sets non-boolean options. This way, we test that the values that we're proposing are sensible.
* Add an `all.sh` component with `config.pl full` and Clang+Asan. Clang+Asan catches things that GCC+Asan doesn't.
* Fix undefined behavior in HAVEGE. Fix #2598 
* Fix various errors in tests.
* When building with `-m32`, also link with `-m32`, as discussed in https://github.com/ARMmbed/mbedtls/pull/1622#discussion_r290718810
* Fix #2682 `programs` getting rebuilt over and over.

This PR requires the crypto submodule to have the changes from https://github.com/ARMmbed/mbed-crypto/pull/144 . To pass CI, I include a commit that updates the submodule accordingly.

Needs backports:

* `all.sh` changes: except concerning PSA which didn't exist.
* HAVEGE: TBD — compatibility break vs theoretical undefined behavior. But if we don't backport it, we need to disable HAVEGE in the Clang+Asan run.
* `config.pl full` behavior: yes, for uniformity between the branches.
* Test framework changes and support for `MBEDTLS_CHECK_PARAMS` as assert: to 2.16.
* Test code changes: none, they're all in code that uses PSA.

### CODE REVIEW

File: include/mbedtls/config.h [hunk 1]
```diff
@@ -1701,6 +1740,37 @@
...
+ * through mbedtls_ssl_conf_cert_profile. Turning on this option is not
+ * recommended because of it is possible to generate SHA-1 collisions, however
+ * this may be safe for legacy infrastructure where additional controls apply.
+ *
+ * \warning   SHA-1 is considered a weak message digest and its use constitutes
+ *            a security risk. If possible, we recommend avoiding dependencies
+ *            on it, and considering stronger message digests instead.
+ *
+ */
+// #define MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_CERTIFICATES
```
Comment (Patater) [inline include/mbedtls/config.h]: This commit ("Move config.h boolean options to their appropriate section") is doing more than just moving options around. It has also explicitly disabled allowing SHA-1 in certs. I think that's a big enough change that it deserves its own commit and a changelog entry.

File: scripts/config.pl [hunk 2]
```diff
@@ -104,6 +104,7 @@
 MBEDTLS_PSA_CRYPTO_SPM
 MBEDTLS_PSA_INJECT_ENTROPY
 MBEDTLS_ECP_RESTARTABLE
+MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_CERTIFICATES
```
Comment (Patater) [inline scripts/config.pl]: This should probably change along with the disabling of the option by default, in that same commit.

File: scripts/config.pl [hunk 3]
```diff
@@ -129,7 +130,7 @@
 
 # Things that should be enabled in "full" even if they match @excluded
 my @non_excluded = qw(
-PLATFORM_[A-Z0-9]+_ALT
+PLATFORM_(?!ZEROIZE_)[A-Z0-9]+_ALT
```
Comment (Patater) [inline scripts/config.pl]: The commit message doesn't adequately explain why we mustn't exclude MBEDTLS_PLATFORM_ZEROIZE_ALT from being excluded. Could you add the reasoning behind the change to the commit message? In particular, why would we want to prevent zeroize from being included in full?

File: tests/suites/test_suite_x509write.function [hunk 4]
```diff
@@ -39,25 +39,29 @@ static int x509_crt_verifycsr( const unsigned char *buf, size_t buflen )
     unsigned char hash[MBEDTLS_MD_MAX_SIZE];
     const mbedtls_md_info_t *md_info;
     mbedtls_x509_csr csr;
+    int ret;
+
+    mbedtls_x509_csr_init( &csr );
```
Comment (Patater) [inline tests/suites/test_suite_x509write.function]: It's not required to call init on csr before calling csr_parse() and friends, per our documentation, correct? Why add it here?

File: tests/suites/test_suite_x509write.function [hunk 5]
```diff
@@ -39,25 +39,29 @@ static int x509_crt_verifycsr( const unsigned char *buf, size_t buflen )
...
 
-    if( mbedtls_pk_verify_ext( csr.sig_pk, csr.sig_opts, &csr.pk,
-                       csr.sig_md, hash, mbedtls_md_get_size( md_info ),
-                       csr.sig.p, csr.sig.len ) != 0 )
-    {
-        return( MBEDTLS_ERR_X509_CERT_VERIFY_FAILED );
-    }
+    ret = mbedtls_pk_verify_ext( csr.sig_pk, csr.sig_opts, &csr.pk,
+                                 csr.sig_md, hash, mbedtls_md_get_size( md_info ),
+                                 csr.sig.p, csr.sig.len );
```
Comment (Patater) [inline tests/suites/test_suite_x509write.function]: Previously, we'd always return MBEDTLS_ERR_X509_CERT_VERIFY_FAILED no matter what non-zero value we got back from mbedtls_pk_verify_ext(). Now, we can return errors other than MBEDTLS_ERR_X509_CERT_VERIFY_FAILED when mbedtls_pk_verify_ext() can return those. Does this make our tests less robust? Were any of our tests expecting this exact error code and are now more fragile because there may come a time when they don't get that error?

File: tests/suites/test_suite_x509write.function [hunk 6]
```diff
@@ -39,25 +39,29 @@ static int x509_crt_verifycsr( const unsigned char *buf, size_t buflen )
...
+    ret = mbedtls_x509_csr_parse( &csr, buf, buflen );
+    if( ret != 0 )
+        goto exit;
 
     md_info = mbedtls_md_info_from_type( csr.sig_md );
     if( mbedtls_md( md_info, csr.cri.p, csr.cri.len, hash ) != 0 )
     {
         /* Note: this can't happen except after an internal error */
-        return( MBEDTLS_ERR_X509_BAD_INPUT_DATA );
+        ret = MBEDTLS_ERR_X509_BAD_INPUT_DATA;
```
Comment (Patater) [inline tests/suites/test_suite_x509write.function]: Why do we care to return this particular error as opposed to just returning ret as is?

File: include/mbedtls/config.h (hunk 1 above)
Comment (gilles-peskine-arm) [reply include/mbedtls/config.h]: If it had done that, that would be wrong. I do not intend to change any default setting in this PR. But I don't see the problem. This commit only reorders lines.
```
diff <(git show c00d4a138b6543cfceee402a175aae21d2e92ceb~1:include/mbedtls/config.h |sort) <(git show c00d4a138b6543cfceee402a175aae21d2e92ceb:include/mbedtls/config.h |sort)
```

File: scripts/config.pl (hunk 2 above)
Comment (gilles-peskine-arm) [reply scripts/config.pl]: This option has been disabled by default since its inception. Before, it was in a section that `config.pl` didn't touch, so it didn't need to be mentioned in `config.pl`. Now that `config.pl full` processes options in the whole file, this one needs to be excluded explicitly to avoid enabling it (which would be bad because it's a security risk).

File: tests/suites/test_suite_x509write.function (hunk 4 above)
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_x509write.function]: It's required to call `mbedtls_x509_csr_free` in the cleanup code, which was added to fix a memory leak. Since `mbedtls_x509_csr_parse` is the very first thing that this function calls that might fail, the call to `mbedtls_x509_csr_init` is not necessary, but omitting it would be fragile: adding anything that can fail before `mbedtls_x509_csr_parse` would cause `mbedtls_x509_csr_free` to access uninitialized memory if that thing fails.

File: tests/suites/test_suite_x509write.function (hunk 5 above)
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_x509write.function]: `x509_crt_verifycsr` is only used once, in a call `TEST_ASSERT( x509_crt_verifycsr( … ) == 0 )`. So at the moment it doesn't make any difference.

File: tests/suites/test_suite_x509write.function (hunk 6 above)
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_x509write.function]: I suspect this was copy-pasted from an old version of the X509 module, and I don't think this is right: it could be a hardware failure in an MD accelerator, for example. But here in test code it doesn't really matter, and it's outside the scope of this PR so I preferred to leave it alone.

File: include/mbedtls/config.h (hunk 1 above)
Comment (Patater) [reply include/mbedtls/config.h]: I misread as "MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_KEY_EXCHANGE", sorry.

File: tests/suites/test_suite_x509write.function (hunk 5 above)
Comment (Patater) [reply tests/suites/test_suite_x509write.function]: OK, I was concerned about the inconsistent approach to keeping or not the errors returned by other functions between the changes around `mbedtls_md()` and `mbedtls_pk_verify_ext()`.

File: tests/suites/test_suite_x509write.function (hunk 6 above)
Comment (Patater) [reply tests/suites/test_suite_x509write.function]: But the call to `mbedtls_pk_verify_ext()` was not left alone, hence why I was asking. Not that it matters much.

**Comment by gilles-peskine-arm** [comment]
> does #1622 pass all.sh testing?

It took a few tries but it does now.

File: scripts/config.pl (hunk 3 above)
Comment (gilles-peskine-arm) [reply scripts/config.pl]: I added the reasoning in a comment in a subsequent commit. I'm not going to rewrite the old commit because it adhered to the standard for documentation at the time.

File: scripts/config.pl [hunk 7]
```diff
@@ -65,10 +39,19 @@
                               not. The value of the symbol is output if one is
                               specified in the configuration file.
     full                    - Uncomments all #define's in the configuration file
-                              excluding some reserved symbols, until the
-                              'Module configuration options' section
+                              excluding some reserved symbols and the
+                              'Module configuration options' section.
+                              This is a good basis for testing the exclusion
+                              of specific additional options.
+    veryfull                - Uncomments all #define's in the configuration file
```
Comment (sbutcher-arm) [inline scripts/config.pl]: `veryfull` isn't a very descriptive name. Surely we can do better?

Comment (gilles-peskine-arm) [reply scripts/config.pl]: With `full` and `realfull` I thought we were following the telescope naming convention... Better suggestions accepted. And let's rename `realfull` while we're at it.

**Comment by gilles-peskine-arm** [comment]
My original approach to fix https://github.com/ARMmbed/mbedtls/issues/2671 and unblock https://github.com/ARMmbed/mbedtls/pull/1622 didn't actually fix it. This PR contains some good testing improvements, but it's become a bit bloated and meandering. I'm going to split it into more manageable pieces. The first piece is https://github.com/ARMmbed/mbedtls/pull/2697. I intend to base full-config improvements on https://github.com/ARMmbed/mbedtls/pull/2469.
